### ISSUE TITLE
Could anyone explain how to make server pre rendering?

### ISSUE THREAD

**Issue by oleksandrbazil** [DavideViolante/Angular-Full-Stack#75]
How to implement '@angular/platform-browser' module for page pre rendering by server? This functionality need for google bot fetching and social sharing.

**Comment by DavideViolante** [comment]
I think you are talking about Angular Universal: https://github.com/angular/universal
I never tried that.

**Comment by RonAlmog** [comment]
another solution is a website called prerender.io . this is a paid and fairly expensive service (free up to some level) that generates the pages for you. also check this one: www.prerender.cloud.
Google is supposed to be able to read angular pages. after all, they created it, right?  But in fact, they don't. so you better help them read your pages. my first bet today would be universal. give it a try and tell us how it went.

**Comment by oleksandrbazil** [comment]
Thanks @RonAlmog and @DavideViolante. I thought Angular 4.0 has all required functionality to build server prerendering because [Angular Universal](https://github.com/angular/universal) had merged to Angular 4.0. I tried to it, but still can't do it alone. 
If someone find a way to make server prerendering please share examples. 
@DavideViolante  thank you for a nice repository.

**Comment by westor21** [comment]
There is a nice howto here:
https://coursetro.com/posts/code/68/Make-your-Angular-App-SEO-Friendly-(Angular-4-+-Universal)

I tryed the describes steps.
Instead of /src/app/app.server.module.ts I dropped the app.server.module into client/app.
I was also not clear with the server part. Can I use the existing server with it's ajax routes or should I create another one? How does the infrasctructure work together?
I tryed to extend the existing server/app.ts
(line 7ff):
import 'reflect-metadata';
import 'zone.js/dist/zone-node';
import { platformServer, renderModuleFactory } from '@angular/platform-server';
import { enableProdMode } from '@angular/core';
import { AppServerModuleNgFactory } from '../dist/ngfactory/src/app/app.server.module.ngfactory';
import { readFileSync } from 'fs';
import { join } from 'path';

then somewhere leater:

enableProdMode();
const template = readFileSync(join(__dirname, '..', 'dist', 'index.html')).toString();

app.engine('html', (_, options, callback) => {
  const opts = { document: template, url: options.req.url };

  renderModuleFactory(AppServerModuleNgFactory, opts)
    .then(html => callback(null, html));
});

app.set('view engine', 'html');
app.set('views', 'src');

app.get('*.*', express.static(join(__dirname, '..', 'dist')));

app.get('*', (req, res) => {
  res.render('index', { req });
});

I tryed "ng build --prod && ngc"

But that failes with
Error at C:/dev/Angular-Full-Stack/server/app.ts:12:42: Cannot find module '../dist/ngfactory/src/app/app.server.module.ngfactory'.

@DavideViolante - Thank you for your great work. I think, it should not be a big problem for you, to find a working solution. I'm using your project as base for the ajax part at my website in a productive environment - https://www.urlaubszeit.de/aida-kreuzfahrten - but now I will need more SEO friendly linking. Therefore angular universal seems to be what I am looking for.
